File: core/bindings/python/auki-sdk-py/examples/compute_task.py
"""Run an already-provisioned compute node against explicitly configured services.

Live operation: registers /example/uppercase/v1, claims tasks, reads input data,
and creates result data. Provision/submit only within an approved test scope.
Task meta must contain input_id. The submitter owns result cleanup.
"""
import asyncio
import os
import signal

import auki_sdk

CAPABILITY = "/example/uppercase/v1"


async def uppercase(task):
    data = task.data()
    await task.progress({"phase": "reading"})
    source = await data.read(task.meta["input_id"])
    # This example handles small buffers. Use streaming for large artifacts and
    # a cooperative stop mechanism for CPU-heavy work, threads or hardware.
    result = source.decode("utf-8").upper().encode("utf-8")
    await task.progress({"phase": "writing"})
    stored = await data.write(result, name=f"uppercase-{task.id}", data_type="example.text.v1")
    return {"output_cids": [stored["id"]], "meta": {"data_id": stored["id"]}}


async def main():
    credential = auki_sdk.AukiComputeCredential(
        dds_url=os.environ["DDS_BASE_URL"],
        dms_url=os.environ["DMS_BASE_URL"],
        registration=os.environ["NODE_REGISTRATION_CREDENTIAL"],
        wallet_key=os.environ["NODE_WALLET_KEY"],
        version="1.0.0",
        client_id=os.environ["AUKI_CLIENT_ID"],
        peer_identity_file=os.environ.get("AUKI_PEER_IDENTITY_FILE"),
    )
    tasks = auki_sdk.AukiDmsTasks(credential, {CAPABILITY: uppercase})
    running = asyncio.ensure_future(tasks.run())
    loop = asyncio.get_running_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, running.cancel)
    try:
        await running
    except asyncio.CancelledError:
        pass
    finally:
        try:
            # Wait for the actual handler's finally blocks, data operations and
            # registration shutdown even when the run awaitable was cancelled.
            await tasks.close()
        finally:
            await credential.close()
            for sig in (signal.SIGINT, signal.SIGTERM):
                loop.remove_signal_handler(sig)


if __name__ == "__main__":
    asyncio.run(main())

File: core/auki-sdk/examples/compute_task.rs
//! Claim at most one task for an already-provisioned compute node.
//! Live operation: registers, claims work and creates a Domain artifact.
use async_trait::async_trait;
use auki_sdk::{
    AukiComputeCredential, AukiDmsTasks, ComputeConfig, DataWrite, SecretString, TaskContext,
    TaskError, TaskHandler, TaskResult, TasksConfig,
};
use tokio_util::sync::CancellationToken;

struct Uppercase;
#[async_trait]
impl TaskHandler for Uppercase {
    async fn run(&self, task: TaskContext) -> Result<TaskResult, TaskError> {
        let id = task.task.meta["input_id"]
            .as_str()
            .ok_or(TaskError::Handler)?
            .parse()
            .map_err(|_| TaskError::Handler)?;
        let data = task.data();
        task.progress(serde_json::json!({"phase":"reading"}))?;
        let input = data.read(id).await.map_err(|_| TaskError::Handler)?;
        let output = String::from_utf8(input)
            .map_err(|_| TaskError::Handler)?
            .to_uppercase();
        task.progress(serde_json::json!({"phase":"writing"}))?;
        let name = format!("uppercase-{}", task.task.id);
        let stored = data
            .write(
                DataWrite::Named {
                    name: &name,
                    data_type: "example.text.v1",
                },
                output.as_bytes(),
            )
            .await
            .map_err(|_| TaskError::Handler)?;
        Ok(TaskResult {
            output_cids: vec![stored.id.to_string()],
            meta: serde_json::json!({"data_id":stored.id}),
        })
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let config = ComputeConfig::new(
        &std::env::var("DDS_BASE_URL")?,
        &std::env::var("DMS_BASE_URL")?,
        SecretString::new(std::env::var("NODE_REGISTRATION_CREDENTIAL")?),
        SecretString::new(std::env::var("NODE_WALLET_KEY")?),
        "1.0.0",
        &std::env::var("AUKI_CLIENT_ID")?,
    )?;
    let credential = AukiComputeCredential::new(config)?;
    let tasks = AukiDmsTasks::new(
        credential.clone(),
        vec!["/example/uppercase/v1".into()],
        TasksConfig::default(),
    )?;
    let outcome = tasks
        .run_once(
            "/example/uppercase/v1",
            &Uppercase,
            &CancellationToken::new(),
        )
        .await;
    let cleanup = tasks.close().await;
    credential.close().await;
    cleanup?;
    println!("{:?}", outcome?);
    Ok(())
}
